subprojects/core/src/main/java/org/gradle/profile/ProfileEventAdapter.java

```diff
@@ -23,21 +23,28 @@
 import org.gradle.api.Task;
 import org.gradle.api.artifacts.DependencyResolutionListener;
 import org.gradle.api.artifacts.ResolvableDependencies;
+import org.gradle.api.artifacts.component.ComponentArtifactIdentifier;
 import org.gradle.api.execution.TaskExecutionListener;
 import org.gradle.api.initialization.Settings;
+import org.gradle.api.internal.artifacts.transform.ArtifactTransformListener;
+import org.gradle.api.internal.artifacts.transform.ArtifactTransformer;
 import org.gradle.api.invocation.Gradle;
 import org.gradle.api.tasks.TaskState;
 import org.gradle.initialization.BuildCompletionListener;
 import org.gradle.internal.buildevents.BuildStartedTime;
 import org.gradle.internal.time.Clock;
 
+import javax.annotation.Nullable;
+import java.io.File;
+
 /**
  * Adapts various events to build a {@link BuildProfile} model, and then notifies a {@link ReportGeneratingProfileListener} when the model is ready.
  */
-public class ProfileEventAdapter implements BuildListener, ProjectEvaluationListener, TaskExecutionListener, DependencyResolutionListener, BuildCompletionListener {
+public class ProfileEventAdapter implements BuildListener, ProjectEvaluationListener, TaskExecutionListener, DependencyResolutionListener, BuildCompletionListener, ArtifactTransformListener {
     private final BuildStartedTime buildStartedTime;
     private final Clock clock;
     private final ProfileListener listener;
+    private final ThreadLocal<ContinuousOperation> currentTransform = new ThreadLocal<ContinuousOperation>();
```

Comment:
Why are we using a `ThreadLocal` to track the current transform and don't get the transform from the build profile after it ran like we do for tasks and dependencies?

I don't think the listener runs on the same thread as the one executing the artifact transform. So I guess this may even give the wrong answer for artifact transforms running in parallel.

Reply:
They run on the same thread: https://github.com/gradle/gradle/blob/0462f98d29009435ad627eff12a022cb9f712e9d/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformArtifactOperation.java#L51-L60

We use this ugly trick because an artifact is still reported to be transformed several times (even though in all but one case it is already cached), and we add up the time spent like this.

Reply:
I know that the transform runs on the same thread. But the listener notification may run on another thread. See https://github.com/gradle/gradle/blob/0c4a1ab113654969fd2679e9e201b4ef75daad71/subprojects/messaging/src/main/java/org/gradle/internal/event/ListenerManager.java#L58.